Define byte

ehuss opened the pull request:
This is split off from https://github.com/rust-lang/reference/pull/1664 which defines what a "byte" is.

Closes https://github.com/rust-lang/reference/issues/1673

cc @rust-lang/opsem, would you be willing to review?

ehuss committed:
Define byte

This is split off from https://github.com/rust-lang/reference/pull/1664
which defines what a "byte" is.

Closes https://github.com/rust-lang/reference/issues/1673

RalfJung commented on src/memory-model.md:
```diff
@@ -1,5 +1,26 @@
+The most basic unit of memory in Rust is a byte. All values are computed from zero or more bytes read from an allocation.
+
+> [!NOTE]
+> While bytes are typically lowered to hardware bytes, Rust uses an "abstract" notion of bytes that can make distinctions which are absent in hardware, such as being uninitialized, or storing part of a pointer.
```
```suggestion
> While bytes are typically lowered to hardware bytes, Rust uses an "abstract" notion of bytes that can make distinctions which are absent in hardware, such as being uninitialized, or storing part of a pointer. Those distinctions can affect whether your program has undefined behavior, so they still have tangible impact on how compiled Rust programs behave!
```

RalfJung commented:
This seems fairly non-controversial, given in particular that we do not say what exactly a "provenance" looks like.

However, it does seem to me like we should involve @rust-lang/lang as well, given that this will make stable promises. Though I cannot entirely articulate what we are actually committing to -- I guess it is the fact that uninitialized memory is "real" as far as Rust is concerned, and distinct from all 256 regular initialized states of a byte. But that's long-established consensus, isn't it?

RalfJung commented on src/memory-model.md:
```diff
@@ -1,5 +1,26 @@
+r[memory.bytes.contents]
+Each byte may have one of the following values:
+
+r[memory.bytes.init]
+* An initialized byte containing a `u8` value and optional [provenance][std::ptr#provenance],
+
+r[memory.bytes.uninit]
+* An uninitialized byte.
```
I find this enumeration interrupted by `r[...]` quite hard to read. But maybe that can be fixed with better stylesheets? There was some discussion about moving the `r` headers to the left of the text if there is enough horizontal space, is that going to happen?

RalfJung replied:
Ah, I looked at https://doc.rust-lang.org/reference/behavior-considered-undefined.html. The [nightly](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html) version looks a lot better indeed. :)

ehuss committed:
Add suggestion from RalfJung

ehuss committed:
Drop sentence about being from an allocation

We didn't feel like this was well-defined enough to go into right now, and didn't seem necessary to include right now.
